### ISSUE TITLE
Analysis of accessibility considerations

### ISSUE THREAD

**Issue by mejackreed** [ProjectMirador/mirador#1664]
Goal: some sort of checklist of accessibility issues we want to ensure are evaluated as new features are developed. 

Considerations include keyboard accessibility, landmarks, and things in the WCAG guidelines. Details to be discussed among @jvine, @sdellis, and @ggeisler.

**Comment by sdellis** [comment]
In case there was any question about [WCAG 2.1 adoption in Europe](https://www.w3.org/blog/2018/09/wcag-2-1-adoption-in-europe/):

> Last week the three official European Standards Organizations (ESOs), CEN, CENELEC, and ETSI, published an updated version of EN 301 549 “Accessibility requirements for ICT products and services”. The most significant change in this version is adoption of the W3C “Web Content Accessibility Guidelines (WCAG) 2.1″ for web content, electronic documents, and non-web software, such as native mobile applications.

**Comment by ggeisler** [comment]
Thanks for the various links, @sdellis. Just to be clear, @jvine and I are also thinking there are more general/app-specific things beyond the WCAG guidelines that we might discuss, such as:

- Establishing landmarks in the app
- Defining keyboard tabbing order of elements
- Deciding how the user will do things like resize and move companion windows with a keyboard

We discussed making specific tickets for the different a11y things we need to think through and produce some sort of guideline for, such as the items above. So that might be one goal for our meeting later this week: decide what a11y guidelines need to be considered and produced, and make a ticket for each of them.

**Comment by sdellis** [comment]
I think this was lost in a PR... I looked at the MUI markup and attempted to create an ideal markup for semantic HTML and aria landmarks. Putting it here for safe keeping:

```
<div id="mirador-app">
  <div class="mirador-workspace-control-panel">
    <nav id="mirador-workspace-control-panel-menu">
      <ul>
        <li>
          <button type="button" id="addBtn" alt="Add Window to Workspace" aria-haspopup="true">
            <span class="MuiFab-label-50">
              <svg ... />
            </span>
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <main class="mirador-workspace" role="main">
    <section class="mirador-window">
      <header>
        <h1>Title of Work 1</h1>
        <nav aria-label="Window Menu">
        		<ul>
        				<li class="has-submenu">
                  <a aria-haspopup="true" aria-expanded="false">⋮</a>
                  <ul>
                      <li><a href="…">Space Bear 6</a></li>
                      <li><a href="…">Space Bear 6 Plus</a></li>
                  </ul>
                </li>
        				<li><a>Maximize</a></li>
        				<li><a>Close</a></li>
        		</ul>
        </nav>
      </header>
      <div class="mirador-window-control-panel">
        <nav class="mirador-window-control-panel-menu">
          <ul>...</ul>
        </nav>
      </div>
      <div class="mirador-window-companion-drawer">
        <aside class="metadata">...</aside>
        <aside class="annotations">...</aside>
        <aside class="search">...</aside>
      </div>
      <div class="mirador-window-content-container" role="main">
        <div>
          <!-- image that displays while loading -->
          <img src="http://foobar" alt="foobar"/>
        </div>
        <div class="openseadragon">
          ...
        </div>
      </div>
    </section>
    <section class="mirador-window">
      <header><h1>Title of Work 2</h1></header>
      <div class="mirador-window-control-panel">
        <nav class="mirador-window-control-panel-menu">
          <ul>...</ul>
        </nav>
      </div>
      <div class="mirador-window-content-container" role="main">
        <div>
          <!-- image that displays while loading -->
          <img src="http://foobar" alt="foobar"/>
        </div>
        <div class="openseadragon">
          ...
        </div>
      </div>
    </section>
    ...
  </main>
</div>
```

**Comment by jvine** [comment]
Overall structure looks good, @sdellis.

Referring to this article - https://benrobertson.io/accessibility/designing-layouts-for-screen-readers - I played with adding ARIA to make the landmarks more useful. Current structure gives this in VoiceOver:

<img width="493" alt="screen shot 2019-01-30 at 10 07 07 am" src="https://user-images.githubusercontent.com/6945691/52002420-d2a0b880-2476-11e9-87f6-51d092a19094.png">

ARIA could give something like this:

<img width="495" alt="landmarks2" src="https://user-images.githubusercontent.com/6945691/52019933-45745880-24a4-11e9-9713-abb0ed9646a4.png">

Not sure I have the right level of detail; the goal is to provide context and unique labels for landmarks that can exist in multiple windows. 

```
<div id="mirador-app">
  <div class="mirador-workspace-control-panel">
    <nav id="mirador-workspace-control-panel-menu" aria-label="Workspace">
      <ul>
        <li>
          <button type="button" id="addBtn" alt="Add Window to Workspace" aria-haspopup="true">
            Add
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <main class="mirador-workspace" role="main" aria-label="Workspace">
    <section class="mirador-window" aria-label="Window 1">
      <header>
        <h1>Title of Work 1</h1>
        <nav aria-label="Window 1 tools">
          <ul>
            <li class="has-submenu">
              <a aria-haspopup="true" aria-expanded="false">⋮</a>
              <ul>
                <li><a href="…">Space Bear 6</a></li>
                <li><a href="…">Space Bear 6 Plus</a></li>
              </ul>
            </li>
            <li><a>Maximize</a></li>
            <li><a>Close</a></li>
          </ul>
        </nav>
      </header>
      <div class="mirador-window-control-panel">
        <nav class="mirador-window-control-panel-menu" aria-label="Window 1 content">
          <ul><li>Window control</li></ul>
        </nav>
      </div>
      <div class="mirador-window-companion-drawer">
        <aside class="metadata" aria-label="Window 1 metadata">metadata</aside>
        <aside class="annotations" aria-label="Window 1 annotations">annos</aside>
        <aside class="search" aria-label="Window 1 search">search</aside>
      </div>
      <section class="mirador-window-content-container" role="main" aria-label="(Canvas label)">
        <div>
          <!-- image that displays while loading -->
          <img src="http://foobar" alt="foobar">
        </div>
        <div class="openseadragon">OSD here</div>
      </section>
    </section>
    <section class="mirador-window" aria-label="Window 2">
      <header>
        <h1>Title of Work 2</h1>
        <nav aria-label="Window 2 tools">
          <ul>
            <li class="has-submenu">
              <a aria-haspopup="true" aria-expanded="false">⋮</a>
              <ul>
                <li><a href="…">Space Bear 6</a></li>
                <li><a href="…">Space Bear 6 Plus</a></li>
              </ul>
            </li>
            <li><a>Maximize</a></li>
            <li><a>Close</a></li>
          </ul>
        </nav>
      </header>
      <div class="mirador-window-control-panel">
        <nav class="mirador-window-control-panel-menu" aria-label="Window 2 content">
          <ul><li>Window control</li></ul>
        </nav>
      </div>
      <div class="mirador-window-companion-drawer">
        <aside class="metadata" aria-label="Window 2 about">metadata</aside>
        <aside class="annotations" aria-label="Window 2 annotations">annos</aside>
        <aside class="search" aria-label="Window 2 search">search</aside>
      </div>
      <section class="mirador-window-content-container" role="main" aria-label="(Canvas label)">
        <div>
          <!-- image that displays while loading -->
          <img src="http://foobar" alt="foobar">
        </div>
        <div class="openseadragon">OSD here</div>
      </section>
    </section>
    ...
  </main>
</div> 
```

**Comment by sdellis** [comment]
@jvine This is great, and yes, much more helpful. I was anticipating that "Window 1" and "Window 2" would instead be the title of the work. Perhaps it could be "Window 1, Title of Work"?

**Comment by jvine** [comment]
I thought of that as well, but the risk is the landmarks looking like this:

`Window 1, Edwards, Richard. Analytical sixth reader :containing an introductory article on the general principles of elocution, with a thorough method of analysis, intended to develop the pupil's appreciation of the thought and emotion, and a critical phonic analyis of English words : designed for the use of normal and high schools, and the highest classes in common schools. New York : Taintor & Co., c1867. region`

and the title is also present in the `h1` within the region. So I backed off from including it... but then felt like it was missing. Can we verify with someone who uses landmarks how a really long landmark name might affect their use?

**Comment by jvine** [comment]
Spinoff tickets:
#2024 Apply accessible heading structure
#2033 Add accessibility landmarks

**Comment by cbeer** [comment]
Do you expect to spin off additional tickets from this?

**Comment by ggeisler** [comment]
@sdellis You're currently assigned to this but the main concerns of this ticket are done so we're closing this. Other accessibility concerns are represented in other tickets.
